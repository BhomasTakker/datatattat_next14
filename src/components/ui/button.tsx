import { StyleSheet } from "@/types/css";
import styles from "./button.module.scss";
import { ButtonHTMLAttributes } from "react";

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
	children?: React.ReactNode;
	onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
	classes?: StyleSheet;
};

export const Button = ({
	children,
	onClick,
	classes,
	type = "button",
	...rest
}: ButtonProps) => {
	return (
		<button
			className={`${styles.button} ${classes ? classes.button : ""}`}
			onClick={onClick}
			// need use type else defaults
			//to submit when used in a form...
			type={type}
			{...rest}
		>
			{children ? children : null}
		</button>
	);
};
