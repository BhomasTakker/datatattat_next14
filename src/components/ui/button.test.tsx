import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Button } from "./button";

describe("Button component", () => {
	it("renders children correctly", () => {
		const { getByText } = render(<Button>Click me</Button>);
		expect(getByText("Click me")).toBeInTheDocument();
	});

	it("applies additional classes", () => {
		const { getByRole } = render(<Button classes="extra-class">Test</Button>);
		expect(getByRole("button")).toHaveClass("extra-class");
	});

	it("calls onClick handler when clicked", () => {
		const handleClick = jest.fn();
		const { getByRole } = render(<Button onClick={handleClick}>Click</Button>);
		fireEvent.click(getByRole("button"));
		expect(handleClick).toHaveBeenCalledTimes(1);
	});

	it("sets the button type to 'button' by default", () => {
		const { getByRole } = render(<Button>Default Type</Button>);
		expect(getByRole("button")).toHaveAttribute("type", "button");
	});

	it("sets the button type when provided", () => {
		const { getByRole } = render(<Button type="submit">Submit</Button>);
		expect(getByRole("button")).toHaveAttribute("type", "submit");
	});

	it("spreads additional props to the button", () => {
		const { getByRole } = render(<Button aria-label="my-button">Label</Button>);
		expect(getByRole("button")).toHaveAttribute("aria-label", "my-button");
	});

	it("renders nothing for children if not provided", () => {
		const { getByRole } = render(<Button />);
		expect(getByRole("button").textContent).toBe("");
	});
});
