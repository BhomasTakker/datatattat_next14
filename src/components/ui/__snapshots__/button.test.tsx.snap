// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Snaps renders Button unchanged 1`] = `
<div>
  <button
    aria-label="my-button"
    class="button extra-class"
    type="button"
  >
    Click me
  </button>
</div>
`;
